src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs

```diff
@@ -152,16 +153,24 @@ private static HttpConnectionKey GetConnectionKey(HttpRequestMessage request, Ur
                 }
             }
 
+            HttpHeaderValueCollection<ProductHeaderValue> upgradeHeader = request.Headers.Upgrade;
+            bool isWebSocketConnect = upgradeHeader == null ? false : upgradeHeader.ToString() == "websocket";
```

Comment:
>David, what's special about web sockets? Should anything that upgrades always use CONNECT? What's the downside?

Before WebSocket IETF RFC 6455 was finalized, there was a lot of discussion among folks (including industry partners) about how to make WebSocket get thru proxies easily. Many proxies didn't work with WebSocket traffic at the time. This led us (MSFT) and other stacks to always use CONNECT tunneling for both WS and WSS websocket traffic when going thru a proxy.  All current MSFT stacks for WebSocket (.NET Framework, WinHTTP, WinINet etc.) do CONNECT tunneling thru proxies for both non-secure and secure traffic.

In terms of other things like might use the HTTP Upgrade: header, I don't know.  I'm not aware of other things that are regularly using that header.  But for WebSocket traffic, using CONNECT is important.

Even the current RFC 6455 shows examples about using CONNECT tunneling for non-secure WebSocket traffic:

https://tools.ietf.org/html/rfc6455

```
EXAMPLE: For example, if the client uses an HTTP proxy for all
          traffic, then if it was to try to connect to port 80 on server
          example.com, it might send the following lines to the proxy
          server:

              CONNECT example.com:80 HTTP/1.1
              Host: example.com
```

Reply:
>Does the "websocket" comparison need to be case insensitive, or is case sensitive correct?

https://tools.ietf.org/html/rfc6455 section 4.2.1

```
   3.   An |Upgrade| header field containing the value "websocket",
        treated as an ASCII case-insensitive value.

   4.   A |Connection| header field that includes the token "Upgrade",
        treated as an ASCII case-insensitive value.
```